core/src/main/scala/caliban/wrappers/ApolloPersistedQueries.scala

```diff
@@ -1,58 +1,124 @@
+          docVar.await.flatMap {
+            case Some((_, _, Some(doc))) => ZIO.succeed(doc)
+            case Some((hash, _, None))   => f(query).tap(doc => ApolloPersistence.add(hash, doc))
+            case None                    => f(query)
+          }
+    }
+
+  private def validationWrapper(
+    docVar: Promise[Nothing, Option[(String, GraphQLRequest, Option[Document])]]
+  ): ValidationWrapper[ApolloPersistence] =
+    new ValidationWrapper[ApolloPersistence] {
+      override def wrap[R1 <: ApolloPersistence](
+        f: Wrapper.ValidationWrapperInput => ZIO[R1, ValidationError, ExecutionRequest]
+      ): Wrapper.ValidationWrapperInput => ZIO[R1, ValidationError, ExecutionRequest] =
+        (input: ValidationWrapperInput) =>
+          docVar.await.flatMap {
+            case Some((hash, req, _)) =>
+              hashVariables(req.variables).flatMap { variablesHash =>
+                ApolloPersistence.isValidated(hash, variablesHash).flatMap { isCached =>
+                  if (isCached) f(input.copy(skipValidation = true))
+                  else f(input) <* ApolloPersistence.registerValidation(hash, variablesHash)
+                }
+              }
+            case None                 => f(input)
           }
+
+      /**
+       * In order to cache the validation step against different input variables, we need to strip the value while
+       * retaining the type, with the exception of enums since the validation is performed against the enum value
+       */
+      private def hashVariables(variables: Option[Map[String, InputValue]]): UIO[Int] = {
+        def stripVariableValues: InputValue => UIO[InputValue] = {
+          case NullValue                      => ZIO.succeed(NullValue)
+          case IntValue.IntNumber(_)          => ZIO.succeed(IntValue.IntNumber(0))
+          case IntValue.LongNumber(_)         => ZIO.succeed(IntValue.LongNumber(0L))
+          case IntValue.BigIntNumber(_)       => ZIO.succeed(IntValue.BigIntNumber(0))
+          case FloatValue.FloatNumber(_)      => ZIO.succeed(FloatValue.FloatNumber(0f))
+          case FloatValue.DoubleNumber(_)     => ZIO.succeed(FloatValue.DoubleNumber(0d))
+          case FloatValue.BigDecimalNumber(_) => ZIO.succeed(FloatValue.BigDecimalNumber(0))
+          case StringValue(_)                 => ZIO.succeed(StringValue("?"))
+          case BooleanValue(_)                => ZIO.succeed(BooleanValue(true))
+          case EnumValue(v)                   => ZIO.succeed(EnumValue(v))
+          case InputValue.ListValue(l)        => ZIO.foreach(l)(stripVariableValues).map(InputValue.ListValue(_))
+          case InputValue.ObjectValue(o)      =>
+            val sortedMap = TreeMap.empty[String, InputValue] ++ o
+            ZIO.foreach(sortedMap) { case (k, v) => stripVariableValues(v).map(k -> _) }.map(InputValue.ObjectValue(_))
+          case InputValue.VariableValue(_)    => ZIO.succeed(InputValue.VariableValue("?"))
```

Comment:
I couldn't find anything to suggest that for `VariableValue` we perform validation against the actual value, but leaving this comment just in case I missed something

Reply:
I think you might be right. I've identified another issue with this approach (see https://github.com/ghostdogpr/caliban/pull/1557#issuecomment-1445474759). Happy to hear your thoughts

Reply:
I disabled the skipping of `VariablesCoercer.coerceVariables` for APQs and it seems that's all we need to validate the inputs. This step is very quick so we still get most of the benefit from caching the query validation